mir_eval/transcription.py

```diff
@@ -0,0 +1,376 @@
+
+
+Conventions
+-----------
+
+Notes should be provided in the form of an interval array and a pitch array.
+The interval array contains two columns, one for note onsets and the second
+for note offsets (each row represents a single note). The pitch array contains
+one column with the corresponding note pitch values (one value per note),
+represented by their fundamental frequency (f0) in Hertz.
+
+Metrics
+-------
+
+* :func:`mir_eval.transcription.precision_recall_f1`: The precision,
+  recall, and F-measure of the note transcription, where an estimated note is
+  considered correct if its pitch, onset and (optionally) offset are
+  sufficiently close to a reference note
+
+'''
+
+import numpy as np
+import collections
+from . import util
+import warnings
+
+
+def validate(ref_intervals, ref_pitches, est_intervals, est_pitches):
+    """Checks that the input annotations to a metric look like time intervals
+    and a pitch list, and throws helpful errors if not.
+
+    Parameters
+    ----------
+    ref_intervals : np.ndarray, shape=(n,2)
+        Array of reference notes time intervals (onset and offset times)
+    ref_pitches: np.ndarray, shape=(n,)
+        Array of reference pitch values in Hertz
+    est_intervals : np.ndarray, shape=(m,2)
+        Array of estimated notes time intervals (onset and offset times)
+    est_pitches : np.ndarray, shape=(m,)
+        Array of estimated pitch values in Hertz
+    """
+    # If reference or estimated notes are empty, warn
+    if ref_intervals.size == 0:
+        warnings.warn("Reference notes are empty.")
+    if est_intervals.size == 0:
+        warnings.warn("Estimate notes are empty.")
+
+    # Make sure intervals and pitches match in length
+    if not ref_intervals.shape[0] == ref_pitches.shape[0]:
+        raise ValueError('Reference intervals and pitches have different '
+                         'lengths.')
+    if not est_intervals.shape[0] == est_pitches.shape[0]:
+        raise ValueError('Estimate intervals and pitches have different '
+                         'lengths.')
+
+    # Make sure all pitch values are positive
+    if np.min(ref_pitches) <= 0:
+        raise ValueError("Reference contains at least one non-positive pitch "
+                         "value")
+    if np.min(est_pitches) <= 0:
+        raise ValueError("Estimate contains at least one non-positive pitch "
+                         "value")
+
+
+def match_notes(ref_intervals, ref_pitches, est_intervals, est_pitches,
+                onset_tolerance=0.05, pitch_tolerance=50.0, offset_ratio=0.2,
+                offset_min_tolerance=0.05, strict=False):
+    """Compute a maximum matching between reference and estimated notes,
+    subject to onset, pitch and (optionally) offset constraints.
+
+    Given two note sequences represented by ``ref_intervals``, ``ref_pitches``,
+    ``est_intervals`` and ``est_pitches`` (see ``io.load_valued_intervals``),
```

Comment:
Same here - make this a function link

    :func:`mir_eval.io.load_valued_intervals`